### ISSUE TITLE
Failed to create file (err: 32) when local and remote race

### ISSUE THREAD

**Issue by aganea** [fastbuild/fastbuild#402]
We've somehow managed to get the unfamous ERROR_SHARING_VIOLATION several times in a row.
This is what happens:

```
7> Obj: D:\somepath\myproject_unity1.obj
... (later on)
-> Obj: D:\somepath\myproject_unity1.obj <REMOTE: REMOTE-MACHINE>
... (2 sec after)
4> Obj: D:\somepath\myproject_unity1.obj <LOCAL RACE>
... (a min after)
Failed to create file 'D:\somepath\myproject_unity1.obj' (Err: 32)

```
This is the sequence I suspect leads to failure:
1. The remote agent finishes, sends .obj buffer to local
2. The local worker receives the .obj, cancels the local job (cl.exe which correspond to `<LOCAL RACE>`)
-> at that point I suspect cl.exe was writing the .obj locally on disk.
-> TerminateProcess() being async, cl.exe doesn't shut down right away.
3. The worker tries to write locally the .obj buffer received from the remote.
4. The worker fails after iterating for a total of 2 sec in Client::WriteFileToDisk.

Alternatively, the issue *could* also be caused by Windows Defender, but it happened two times in a days, on different build machines.

There are probably several alternatives to fix this:
- Either increase the timer (not suggested)
- WaitForSingleObject on cl.exe termination, if the first open in WriteFileToDisk fails.
- Run the local race compilation in a different folder, and do a copy like when receiving the remote .obj.

Any other ideas? Thanks in advance!

**Comment by fksfks** [comment]
Forgot to mention, we also see this in the log, just after the error above: (maybe unrelated?)

1> Post failure deletion failed for 'D:\somepath\project.dll'
6> Post failure deletion failed for 'D:\somepath\myproject_unity3.obj'
11>Post failure deletion failed for 'D:\somepath\myproject_unity2.obj'

**Comment by fksfks** [comment]
We ended up doing several things:

In FileIO.cpp, change:
`    WorkAroundForWindowsFilePermissionProblem( const AString & fileName, const uint32_t openMode = FileStream::READ_ONLY, const uint32_t timeoutSeconds = 1 )`
and,
```     while ( f.Open( fileName.Get(), openMode ) == false ). ```
and,
```     if ( timer.GetElapsed() > timeoutSeconds ) ```
accordingly.

In ObjectNode.cpp, instead of :
```
        job->Error( "Failed to create temp file '%s' to build '%s' (error %u)", tmpFileName.Get(), GetName().Get(), Env::GetLastErr() );
        job->OnSystemError();
        return NODE_RESULT_FAILED;
```
we do:
```
        FileIO::WorkAroundForWindowsFilePermissionProblem( tmpFileName, FileStream::WRITE_ONLY, 10 );

        if ( WorkerThread::CreateTempFile(tmpFileName, tmpFile) == false )
        {
            job->Error( "Failed to create temp file '%s' to build '%s' (error %u)", tmpFileName.Get(), GetName().Get(), Env::GetLastErr() );
            job->OnSystemError();
            return NODE_RESULT_FAILED;
        }
```

And in Client.cpp, instead of:
```
        // On Windows, we can occasionally fail to open the file with error 1224 (ERROR_USER_MAPPED_FILE), due to
        // things like anti-virus etc. Simply retry if that happens
        FileIO::WorkAroundForWindowsFilePermissionProblem( fileName );
```
we do:
```
        // When a <LOCAL RACE> occurs, the local compilation process might not have exited at this point (even if we called ::TerminateProcess, which is async),
        // which can cause failure below, because the file is still locked.
        FileIO::WorkAroundForWindowsFilePermissionProblem( fileName, FileStream::WRITE_ONLY, 15 /*seconds max wait time*/ );
```

We haven't seen these errors since then.

**Comment by ffulin** [comment]
This is along the lines of what I was thinking. Thanks for letting me know it solves the issue as I'd hoped.

**Comment by ffulin** [comment]
I've gone ahead and applied your changes to the dev branch (https://github.com/fastbuild/fastbuild/commit/b663325107464cc1e9038ed57bf370dac0a21258). These changes will be in the next release (v0.96). I'll leave this issue open until that version is released. Thanks!

**Comment by ffulin** [comment]
A version containing a fix for this issue has now been released (v0.96).
